2.0

robashton opened the pull request.

adrianroe committed:
Start of 2.0

adrianroe committed:
WIP transfer

adrianroe committed:
WIP

adrianroe committed:
Rework of supervisor types

adrianroe committed:
WIP

adrianroe committed:
WIP

philipstears committed:
WiP gen server start

philipstears committed:
Named Server Test

philipstears committed:
Handle info

adrianroe committed:
gen_server:call implemented

adrianroe committed:
Test for and fix to handle_info breaking outer state

adrianroe committed:
handle_cast and a passing test for it

adrianroe committed:
Style of code in init now matches call and cast

adrianroe committed:
Our first passing supervisor test

adrianroe committed:
Call result typed over reply cont stop msg and state

adrianroe committed:
Example of standalone gen server - ValueServer

robashton commented on Makefile:
```diff
@@ -1,34 +1,19 @@
-
-.psc-package: psc-package.json
-	psc-package install
-	touch .psc-package
```
I put a lot of effort into writing the doc tags or whatever they're called for Pinto (editor support, docs into pursuit and everything). 

I would like to request very loudly that you re-instate this work when you're done.

robashton replied:
Also https://github.com/robashton/purerl-cookbook :)

nwolverson replied:
Generally I wouldn't generate docs into `docs/` unless you're using that for something else, we should ensure that is updated on pursuit (fork) - generally all PureScript libraries used to have generated docs in the repo and moved to rm those and only host on pursuit - with the `purs docs` format outputs just for local/offline doc generation

robashton replied:
@nwolverson Oh sure, I was only doing that to test that I'd done it proper and that it was actually committed for some reason was entirely by the by

robashton commented on test/Main.purs:
```diff
@@ -0,0 +1,226 @@
+
+    where
+      init = do
+        pure $ Right $ InitOk $ TestState 0
```
Right *and* InitOk looks a little odd...

robashton commented on src/Pinto/GenServer.purs:
```diff
@@ -0,0 +1,215 @@
+type InitResult cont state = Either ServerNotRunning (ServerRunning cont state )
+
+-- Can't do a functor instance over a type synonym, so just have a function instead
+mapInitResult :: forall state state' cont. (state -> state') -> InitResult cont state -> InitResult cont state'
```
I can understand wanting to capture "success" vs "failure" in the init results, but the introduction of Either on top of an ADT doesn't really add anything unless you're using 'either' functionality somewhere *on* that result - @nwolverson ?

nwolverson replied:
Yep that looks odd - if the ADT is representing init results it seems like it's just one thing, and stick a function `:: InitResult -> Either _ state` if required on top of that

adrianroe committed:
Via test created and then expunged!

adrianroe committed:
GenServer call functions have access to From

adrianroe committed:
Continue and Reply both implemented and tested

adrianroe committed:
Rework GS return types based on Action, reply and Maybe

adrianroe committed:
WIP on stop normal tests

adrianroe committed:
Should only need a sleep in shutdown tests after a cast initiated stop

robashton commented on test/GenServer.erl:
```diff
@@ -0,0 +1,23 @@
+      code:add_patha("/Users/adrianroe/dev/rtsv2/_build/default/lib/gproc/ebin/"),
+      application:ensure_started(gproc),
+      unit
+  end.
```
You can presumably now use the actual rebar3 eunit test runner with before/after actions to get this sort of thing working properly

adrianroe committed:
Use (ServerType cont stop msg state) to wrap up server type for APIs and supervisors

adrianroe committed:
Emphasise that ServerType is not constructable

robashton commented:
I know the monitor implementation wasn't the greatest in the world, but I'm using that quite heavily in other projects - are you going to replace it?

robashton commented:
Ditto Timer - that went through a lot of iterations and the solution ended up having to be a pragmatic one - that'll need to come back in some form also

philipstears commented:
Yup, they are coming back, we just got rid of them temporarily whilst we were experimenting with replumbing various types.

philipstears committed:
Noodling on representation of GenStatem

philipstears committed:
Ignore .spago directory

philipstears committed:
More noodling on gen_statem types

philipstears committed:
More complete statem type

philipstears committed:
Even more complete statem type

philipstears committed:
handleEnter typing

philipstears committed:
Call typing

philipstears committed:
Sort out some imports

robashton commented on test/DoorLock.purs:
```diff
@@ -0,0 +1,87 @@
+
+    handleEnter UnlockedOpen UnlockedClosed currentData = do
+      audit AuditDoorClosed
+      pure $ Right $ KeepStateAndData
```
Again, this whole Right/Left thing around our ADTs doesn't look right

philipstears replied:
This is an older commit, the Right/Left distinction here was just copy/pasta, it's not there in the current version, except in the init function.

As to whether it should be there in the init function and in the StartLinkResult, I ended up bowing out of that debate, others had fairly strong opinions there and I was pretty much on the fence about the whole thing.

robashton replied:
:ok:

robashton replied:
er, :ok_hand:  even

philipstears committed:
Implement GenStatem, using a custom monad transform rather than surfacing ReaderT directly

philipstears committed:
Restore Monitor, adjusted to work with changes to ServerPid

philipstears committed:
Restore Timer

philipstears committed:
WiP on removing Types.ServerPid

Remove it in favour of gen_* specific types which can be newtype'd by
specific servers/statems so they can control API user access to stuff
like message plinging.

philipstears committed:
InstanceRef removed in favour of ServerRef/StatemRef/SupervisorRef

philipstears committed:
DynamicSup

robashton commented on test/DoorLock.purs:
```diff
@@ -1,38 +1,119 @@
+  getStateId (Locked _) = StateIdLocked
+  getStateId (UnlockedClosed _) = StateIdUnlockedClosed
+  getStateId (UnlockedOpen _) = StateIdUnlockedOpen
 
```
I think I like this

philipstears committed:
GenStatem monitor

srstrong committed:
Added App.purs, fixed bug on self, added initial Sup.stop method and removed commented-out code

srstrong committed:
Fix warnings under 0.14

srstrong committed:
Further warnings squashed

srstrong committed:
Removed io:format

robashton committed:
Fix up shell.nix to use latest bits

robashton committed:
Upgrade to purescript 0.14

robashton committed:
purge left/right from gen server

robashton committed:
In for a penny, in for a pound - Purty all files

robashton committed:
more minor powerpoint-driven-cosmetics

robashton committed:
Remove HasRawPid/HasProcess from Pinto and use the type classes in Erl.Process and Erl.Process.Raw

robashton committed:
rename the FFI for timer

robashton committed:
use the right ****ing function

robashton committed:
Fix tests with Erl.Process.Raw (HasPid)

robashton committed:
Purge warnings, purge more Either(..) nonsense, rename a few things for demo friendliness (mkFooBar on pure fns is bleh)

nwolverson committed:
Modernize makefile

nwolverson committed:
Fix type mismatch in Sup.stop ffi

robashton committed:
types for server ref/server instance and promotion of that concept from FFI to Purerl

robashton committed:
switch the start_child args in dynamic sups to match underlying API + everything else

robashton committed:
rename supervisor flags to match OTP names

robashton committed:
Move FFI into Purescript (removing wrapped types, removing ADT unpacking in Erlang, removing compiler warnings about unused exports)

robashton committed:
purge monitor code from genstatem for now

robashton committed:
Stripped the FFI for Genstatem to align with GenServer

robashton committed:
got rid of unused wrapped fn definitions

robashton committed:
restore trapExit

robashton committed:
Squash some warnings

robashton committed:
a test would have caught that...

robashton committed:
added terminate callback to genserver

robashton committed:
provide a shutdown reason

srstrong committed:
ResultT becomes a newtype to enable it to implement HasMessage

srstrong committed:
Merge

srstrong committed:
Renamed HasMessage to ReceivesMessage

srstrong committed:
Added procLibStartLink and support for NextEvent as an action and StopNormal

robashton committed:
add  trapExit/terminate support to gen statem

robashton committed:
HasSelf implementation for GenServer

robashton committed:
(re) add missing Supervisor functions...

robashton committed:
Missing functions from dynamic sups

robashton committed:
Docs ands some renames, and remove milliseconds/etc

nwolverson committed:
Correct cross-reference to pinto_supervisor@foreign

nwolverson committed:
Suppress warnings around unused FFI

robashton committed:
use self context for timer

robashton committed:
get rid of (msg -> Effect Unit) from Monitor
